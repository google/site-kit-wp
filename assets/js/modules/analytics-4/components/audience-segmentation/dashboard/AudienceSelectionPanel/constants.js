/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
