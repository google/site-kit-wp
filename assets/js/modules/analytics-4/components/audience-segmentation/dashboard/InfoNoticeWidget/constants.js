/**
 * Constants related to audience segmentation info notices.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * UI key to display audience segmentation info notice.
 *
 * @since 1.135.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_HIDE_UI =
	'audience-segmentation-info-notice-ui';

/**
 * Slug for audience segmentation info notice.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_SLUG = 'audience-segmentation-info-notice';

// Notice slugs: new-visitors, compare-metrics, custom-audiences, purchasers, returning-visitors, compare-new-returning, or compare-cities

/**
 * List of info notices for audiences.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICES = [
	{
		slug: 'new-visitors',
		content: __(
			'The higher the portion of new visitors you have, the more your audience is growing. Looking at what content brings them to your site may give you insights on how to reach even more people.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-metrics',
		content: __(
			'Select up to three visitor groups to display on the dashboard and easily compare metrics between them.',
			'google-site-kit'
		),
	},
	{
		slug: 'custom-audiences',
		content: __(
			'Configure your own custom audiences in Analytics to gain deeper insights into visitor behavior, for example consider creating an “Existing customers” or “Subscribers” segment, depending on what goals you have for your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'purchasers',
		content: __(
			'Select the Purchasers visitor group to gain insights into which visitors bring the most revenue to your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'returning-visitors',
		content: __(
			'The more returning visitors your site has, the stronger and more loyal an audience you’re building. Check which content brings people back to your site - it might help you create a strategy to build a community.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-new-returning',
		content: __(
			'Compare the ratio of “new” to “returning” visitors – this can give you insights on whether you have more people stopping by as a one-off, or more loyal visitors.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-cities',
		content: __(
			'Check the cities which bring you more new vs more returning visitors – there might be new audiences you could engage with in locations you hadn’t thought about.',
			'google-site-kit'
		),
	},
];
