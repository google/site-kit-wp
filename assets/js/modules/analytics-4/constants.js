/**
 * GA4 module constants.
 *
 * Site Kit by Google, Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const ACTIVATION_STEP_REMINDER = 0;
export const ACTIVATION_STEP_SETUP = 1;
export const ACTIVATION_STEP_SUCCESS = 2;
export const ACTIVATION_STEP_IN_PROGRESS = 3;

export const ACTIVATION_ACKNOWLEDGEMENT_TOOLTIP_STATE_KEY =
	'activation-acknowledgement-tooltip-state';

export const GA4_ACTIVATION_BANNER_STATE_KEY = 'ga4-activation-banner-state';

export const ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY =
	'enhanced-measurement-activation-banner-tooltip-state';
export const LEGACY_ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY =
	'enhanced-measurement-activation-banner-dismissed-item';

export const REPORT_ARGS_SELECTED_METRIC_KEY = '_r.explorerCard..selmet';
export const REPORT_ARGS_SELECTED_DIMENSION_KEY = '_r.explorerCard..seldim';
export const REPORT_ARGS_DATA_FILTERS_KEY = '_r..dataFilters';
export const REPORT_ARGS_NAV_KEY = '_r..nav';

export const KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY =
	'key-metrics-connect-ga4-cta-widget';

/**
 * Date that Site Kit will automatically switch to using GA4: September 25, 2023.
 *
 * With Site Kit's maximum date range being three months, once this period has passed since the UA cutoff date,
 * it will no longer be able to show any historical data for UA and will automatically switch to GA4.
 * For operational reasons, we automatically make the switch a few days before the three month cutoff point.
 *
 * @since 1.107.0
 * @see UA_CUTOFF_DATE.
 */
export const GA4_AUTO_SWITCH_DATE = '2023-09-25';
export const MODULE_SLUG_ANALYTICS_4 = 'analytics-4';
