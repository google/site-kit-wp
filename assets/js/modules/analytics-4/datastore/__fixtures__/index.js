/**
 * GA4 Datastore Fixtures.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as accountSummaries } from './account-summaries.json';
export { default as audiences } from './audiences.json';
export { default as availableAudiences } from './available-audiences.json';
export { default as container } from './container.json';
export { default as containerE2E } from './container-e2e.json';
export { default as containerDestinations } from './container-destinations.json';
export { default as conversionEvents } from './conversion-events.json';
export { default as createProperty } from './create-property.json';
export { default as createWebDataStream } from './create-webdatastream.json';
export { default as defaultSettings } from './default-settings.json';
export { default as defaultEnhancedMeasurementSettings } from './default-enhanced-measurement-settings.json';
export { default as pageTitles } from './page-titles.json';
export { default as properties } from './properties.json';
export { default as report } from './report.json';
export { default as pivotReport } from './pivot-report.json';
export { default as webDataStreams } from './webdatastreams.json';
export { default as webDataStreamsBatch } from './webdatastreams-batch.json';
export { default as webDataStreamsBatchSetup } from './webdatastreams-batch-setup.json';
export { default as googleTagSettings } from './google-tag-settings.json';
