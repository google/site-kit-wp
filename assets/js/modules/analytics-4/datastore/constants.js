/**
 * `modules/analytics-4` data store constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const MODULES_ANALYTICS_4 = 'modules/analytics-4';

// A special Account ID value used for the "Set up a new account" option.
export const ACCOUNT_CREATE = 'account_create';

// A special Property ID value used for the "Set up a new property" option.
export const PROPERTY_CREATE = 'property_create';

export const WEBDATASTREAM_CREATE = 'webdatastream_create';

export const FORM_SETUP = 'analyticsSetup';

export const MAX_WEBDATASTREAMS_PER_BATCH = 10;

// Date range offset days for Analytics 4 report requests.
export const DATE_RANGE_OFFSET = 1;

export const GTM_SCOPE = 'https://www.googleapis.com/auth/tagmanager.readonly';

export const ENHANCED_MEASUREMENT_FORM = 'enhanced-measurement-form';
export const ENHANCED_MEASUREMENT_ENABLED = 'enhanced-measurement-enabled';
export const ENHANCED_MEASUREMENT_SHOULD_DISMISS_ACTIVATION_BANNER =
	'enhanced-measurement-should-dismiss-activation-banner';

// Form ID for the account creation form.
export const FORM_ACCOUNT_CREATE = 'analyticsAccountCreate';

export const FORM_CUSTOM_DIMENSIONS_CREATE = 'analyticsCustomDimensionsCreate';

// OAuth scope required for provisioning a Google Analytics account.
export const PROVISIONING_SCOPE =
	'https://www.googleapis.com/auth/analytics.provision';
export const EDIT_SCOPE = 'https://www.googleapis.com/auth/analytics.edit';

// Dashboard widget constants.
export const UI_DIMENSION_NAME = 'dashboardAllTrafficWidgetDimensionName';
export const UI_DIMENSION_COLOR = 'dashboardAllTrafficWidgetDimensionColor';
export const UI_DIMENSION_VALUE = 'dashboardAllTrafficWidgetDimensionValue';
export const UI_ACTIVE_ROW_INDEX = 'dashboardAllTrafficWidgetActiveRowIndex';
export const UI_ALL_TRAFFIC_LOADED = 'dashboardAllTrafficWidgetLoaded';

// Note: names and descriptions are not translated as these are not surfaced in Site Kit
// and are also subject to hard limits on the length which would be unpredictable if translated.
// See https://developers.google.com/analytics/devguides/config/admin/v1/rest/v1beta/properties.customDimensions#CustomDimension
export const CUSTOM_DIMENSION_DEFINITIONS = {
	googlesitekit_post_date: {
		parameterName: 'googlesitekit_post_date',
		displayName: 'WordPress Post Date',
		description: 'Created by Site Kit: Date when a post was published',
		scope: 'EVENT',
	},
	googlesitekit_post_author: {
		parameterName: 'googlesitekit_post_author',
		displayName: 'WordPress Post Author',
		description: 'Created by Site Kit: WordPress name of the post author',
		scope: 'EVENT',
	},
	googlesitekit_post_categories: {
		parameterName: 'googlesitekit_post_categories',
		displayName: 'WordPress Post Categories',
		description:
			'Created by Site Kit: Names of categories assigned to a post',
		scope: 'EVENT',
	},
	googlesitekit_post_type: {
		parameterName: 'googlesitekit_post_type',
		displayName: 'WordPress Post Type',
		description: 'Created by Site Kit: Content type of a post',
		scope: 'EVENT',
	},
};

export const ENUM_CONVERSION_EVENTS = {
	ADD_TO_CART: 'add_to_cart',
	PURCHASE: 'purchase',
	SUBMIT_LEAD_FORM: 'submit_lead_form',
	GENERATE_LEAD: 'generate_lead',
	CONTACT: 'contact',
};
export const CONVERSION_REPORTING_LEAD_EVENTS = [
	ENUM_CONVERSION_EVENTS.CONTACT,
	ENUM_CONVERSION_EVENTS.GENERATE_LEAD,
	ENUM_CONVERSION_EVENTS.SUBMIT_LEAD_FORM,
];

// Audience enums.
export const AUDIENCE_FILTER_CLAUSE_TYPE_ENUM = {
	AUDIENCE_CLAUSE_TYPE_UNSPECIFIED: 'AUDIENCE_CLAUSE_TYPE_UNSPECIFIED',
	INCLUDE: 'INCLUDE',
	EXCLUDE: 'EXCLUDE',
};

export const AUDIENCE_FILTER_SCOPE_ENUM = {
	AUDIENCE_FILTER_SCOPE_UNSPECIFIED: 'AUDIENCE_FILTER_SCOPE_UNSPECIFIED',
	AUDIENCE_FILTER_SCOPE_WITHIN_SAME_EVENT:
		'AUDIENCE_FILTER_SCOPE_WITHIN_SAME_EVENT',
	AUDIENCE_FILTER_SCOPE_WITHIN_SAME_SESSION:
		'AUDIENCE_FILTER_SCOPE_WITHIN_SAME_SESSION',
	AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS:
		'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
};

export const SITE_KIT_AUDIENCE_DEFINITIONS = {
	'new-visitors': {
		description: 'People who visited the site for the first time',
		displayName: 'New visitors',
		membershipDurationDays: -1, // The longest duration, 540 days.
		filterClauses: [
			{
				clauseType: 'INCLUDE',
				simpleFilter: {
					scope: 'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
					filterExpression: {
						andGroup: {
							filterExpressions: [
								{
									orGroup: {
										filterExpressions: [
											{
												dimensionOrMetricFilter: {
													fieldName: 'newVsReturning',
													stringFilter: {
														matchType: 'EXACT',
														value: 'new',
													},
												},
											},
										],
									},
								},
								{
									orGroup: {
										filterExpressions: [
											{
												notExpression: {
													dimensionOrMetricFilter: {
														fieldName: 'groupId',
														stringFilter: {
															matchType: 'EXACT',
															value: 'created_by_googlesitekit:new_visitors',
														},
													},
												},
											},
										],
									},
								},
							],
						},
					},
				},
			},
		],
	},
	'returning-visitors': {
		description: 'People who have visited your site at least once before',
		displayName: 'Returning visitors',
		membershipDurationDays: -1, // The longest duration, 540 days.
		filterClauses: [
			{
				clauseType: 'INCLUDE',
				simpleFilter: {
					scope: 'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
					filterExpression: {
						andGroup: {
							filterExpressions: [
								{
									orGroup: {
										filterExpressions: [
											{
												dimensionOrMetricFilter: {
													fieldName: 'newVsReturning',
													stringFilter: {
														matchType: 'EXACT',
														value: 'returning',
													},
												},
											},
										],
									},
								},
								{
									orGroup: {
										filterExpressions: [
											{
												notExpression: {
													dimensionOrMetricFilter: {
														fieldName: 'groupId',
														stringFilter: {
															matchType: 'EXACT',
															value: 'created_by_googlesitekit:returning_visitors',
														},
													},
												},
											},
										],
									},
								},
							],
						},
					},
				},
			},
		],
	},
};

export const AUDIENCE_SEGMENTATION_SETUP_FORM = 'audiencePermissionsSetup';

export const AUDIENCE_TILE_CUSTOM_DIMENSION_CREATE =
	'audienceTileCustomDimensionCreate';

export const AUDIENCE_TYPE_SORT_ORDER = {
	USER_AUDIENCE: 0,
	SITE_KIT_AUDIENCE: 1,
	DEFAULT_AUDIENCE: 2,
};

export const AUDIENCE_ITEM_NEW_BADGE_SLUG_PREFIX =
	'audience-selection-panel-expirable-new-badge-';
