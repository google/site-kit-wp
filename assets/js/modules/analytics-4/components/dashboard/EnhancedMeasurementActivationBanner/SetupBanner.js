/**
 * EnhancedMeasurementActivationBanner > SetupBanner component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { addQueryArgs } from '@wordpress/url';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { CORE_FORMS } from '@/js/googlesitekit/datastore/forms/constants';
import { CORE_LOCATION } from '@/js/googlesitekit/datastore/location/constants';
import { CORE_NOTIFICATIONS } from '@/js/googlesitekit/notifications/datastore/constants';
import { CORE_SITE } from '@/js/googlesitekit/datastore/site/constants';
import { CORE_USER } from '@/js/googlesitekit/datastore/user/constants';
import {
	EDIT_SCOPE,
	FORM_SETUP,
} from '@/js/modules/analytics-4/datastore/constants';
import { ERROR_CODE_MISSING_REQUIRED_SCOPE } from '@/js/util/errors';
import { DAY_IN_SECONDS } from '@/js/util';
import { NOTIFICATION_GROUPS } from '@/js/googlesitekit/notifications/constants';
import SetupCTA from '@/js/googlesitekit/notifications/components/layout/SetupCTA';
import BannerSVG from '@/svg/graphics/banner-enhanced-measurement-setup-cta.svg?url';
import BannerMobileSVG from '@/svg/graphics/banner-enhanced-measurement-setup-cta-mobile.svg?url';

export default function SetupBanner( props ) {
	const { id, Notification, errorNotice, isSaving, onDismiss, onSubmit } =
		props;

	// The `enhanced_measurement` query value is arbitrary and serves two purposes:
	// 1. To ensure that `authentication_success` isn't appended when returning from OAuth.
	// 2. To guarantee it doesn't match any existing notifications in the `BannerNotifications` component, thus preventing any unintended displays.
	const redirectURL = addQueryArgs( global.location.href, {
		notification: 'enhanced_measurement',
	} );

	const hasEditScope = useSelect( ( select ) =>
		select( CORE_USER ).hasScope( EDIT_SCOPE )
	);

	const documentationURL = useSelect( ( select ) =>
		select( CORE_SITE ).getGoogleSupportURL( {
			path: '/analytics/answer/9216061',
		} )
	);

	const isNavigatingToOAuthURL = useSelect( ( select ) => {
		const OAuthURL = select( CORE_USER ).getConnectURL( {
			additionalScopes: [ EDIT_SCOPE ],
			redirectURL,
		} );

		if ( ! OAuthURL ) {
			return false;
		}

		return select( CORE_LOCATION ).isNavigatingTo( OAuthURL );
	} );

	const { setValues } = useDispatch( CORE_FORMS );
	const { setPermissionScopeError } = useDispatch( CORE_USER );
	const { pinNotification } = useDispatch( CORE_NOTIFICATIONS );

	const handleSubmitChanges = useCallback( async () => {
		const scopes = [];

		if ( hasEditScope === false ) {
			scopes.push( EDIT_SCOPE );
		}

		// If scope not granted, trigger scope error right away. These are
		// typically handled automatically based on API responses, but
		// this particular case has some special handling to improve UX.
		if ( scopes.length > 0 ) {
			setValues( FORM_SETUP, { autoSubmit: true } );

			await pinNotification( id, NOTIFICATION_GROUPS.SETUP_CTAS );

			setPermissionScopeError( {
				code: ERROR_CODE_MISSING_REQUIRED_SCOPE,
				message: __(
					'Additional permissions are required to enable enhanced measurement for the selected web data stream',
					'google-site-kit'
				),
				data: {
					status: 403,
					scopes,
					skipModal: true,
					redirectURL,
				},
			} );
			return;
		}

		await onSubmit();
	}, [
		hasEditScope,
		id,
		onSubmit,
		pinNotification,
		redirectURL,
		setPermissionScopeError,
		setValues,
	] );

	const isUsingProxy = useSelect( ( select ) =>
		select( CORE_SITE ).isUsingProxy()
	);
	const { triggerSurvey } = useDispatch( CORE_USER );

	const handleView = useCallback( () => {
		if ( isUsingProxy ) {
			triggerSurvey( 'view_enhanced_measurement_cta', {
				ttl: DAY_IN_SECONDS,
			} );
		}
	}, [ triggerSurvey, isUsingProxy ] );

	const description = hasEditScope
		? __(
				'Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required.',
				'google-site-kit'
		  )
		: __(
				'Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required — you’ll be redirected to give permission for Site Kit to enable it on your behalf.',
				'google-site-kit'
		  );

	return (
		<Notification onView={ handleView }>
			<SetupCTA
				notificationID={ id }
				title={ __(
					'Understand how visitors interact with your content',
					'google-site-kit'
				) }
				description={ description }
				errorText={ errorNotice?.message }
				learnMoreLink={ {
					href: documentationURL,
				} }
				ctaButton={ {
					label: __( 'Enable now', 'google-site-kit' ),
					onClick: handleSubmitChanges,
					disabled: isSaving || isNavigatingToOAuthURL,
					inProgress: isSaving || isNavigatingToOAuthURL,
				} }
				dismissButton={ {
					label: __( 'Maybe later', 'google-site-kit' ),
					onClick: onDismiss,
					disabled: isSaving || isNavigatingToOAuthURL,
				} }
				helpText={ __(
					'You can always add/edit this in the Site Kit Settings',
					'google-site-kit'
				) }
				svg={ {
					desktop: BannerSVG,
					mobile: BannerMobileSVG,
					verticalPosition: 'center',
				} }
			/>
		</Notification>
	);
}

SetupBanner.propTypes = {
	id: PropTypes.string.isRequired,
	Notification: PropTypes.elementType,
	errorNotice: PropTypes.object,
	isSaving: PropTypes.bool,
	onSubmit: PropTypes.func,
	onDismiss: PropTypes.func,
};
