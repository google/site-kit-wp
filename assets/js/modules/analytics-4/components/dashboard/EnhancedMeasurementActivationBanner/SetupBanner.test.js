/**
 * EnhancedMeasurementActivationBanner > SetupBanner component tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
/**
 * Internal dependencies
 */
import {
	createTestRegistry,
	render,
	provideModules,
	provideUserAuthentication,
	provideSiteInfo,
	waitFor,
} from '../../../../../../../tests/js/test-utils';
import { mockSurveyEndpoints } from '../../../../../../../tests/js/mock-survey-endpoints';
import { EDIT_SCOPE, MODULES_ANALYTICS_4 } from '../../../datastore/constants';
import { ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY } from '../../../constants';
import { properties } from '../../../datastore/__fixtures__';
import SetupBanner from './SetupBanner';
import {
	getViewportWidth,
	setViewportWidth,
} from '../../../../../../../tests/js/viewport-width-utils';

describe( 'SetupBanner', () => {
	const propertyID = '1000';
	const webDataStreamID = '2000';

	let enabledSettingsMock;
	let disabledSettingsMock;
	let registry;
	let originalViewportWidth;

	beforeEach( () => {
		enabledSettingsMock = {
			fileDownloadsEnabled: null,
			name: 'properties/1000/dataStreams/2000/enhancedMeasurementSettings',
			outboundClicksEnabled: null,
			pageChangesEnabled: null,
			scrollsEnabled: null,
			searchQueryParameter: 'q,s,search,query,keyword',
			siteSearchEnabled: null,
			streamEnabled: true,
			uriQueryParameter: null,
			videoEngagementEnabled: null,
		};

		disabledSettingsMock = {
			...enabledSettingsMock,
			streamEnabled: false,
		};

		registry = createTestRegistry();

		provideUserAuthentication( registry );
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			propertyID,
			webDataStreamID,
		} );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings( disabledSettingsMock, {
				propertyID,
				webDataStreamID,
			} );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], { propertyID } );

		fetchMock.postOnce(
			RegExp( '^/google-site-kit/v1/core/user/data/dismiss-item' ),
			{
				body: JSON.stringify( [
					ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY,
				] ),
				status: 200,
			}
		);

		originalViewportWidth = getViewportWidth();
		setViewportWidth( 450 );
	} );

	afterEach( () => {
		setViewportWidth( originalViewportWidth );
	} );

	it( 'should render correctly when the user does have the edit scope granted', async () => {
		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		const { container, getByText } = render( <SetupBanner />, {
			registry,
		} );

		await waitFor( () => expect( container ).toMatchSnapshot() );

		await waitFor( () =>
			expect(
				getByText(
					'Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required.'
				)
			).toBeInTheDocument()
		);
	} );

	it( 'should render correctly when the user does not have the edit scope granted', () => {
		mockSurveyEndpoints();

		provideSiteInfo( registry, {
			usingProxy: true,
		} );

		const { container, getByText } = render( <SetupBanner />, {
			registry,
		} );

		expect( container ).toMatchSnapshot();

		expect(
			getByText(
				'Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required — you’ll be redirected to give permission for Site Kit to enable it on your behalf.'
			)
		).toBeInTheDocument();
	} );
} );
