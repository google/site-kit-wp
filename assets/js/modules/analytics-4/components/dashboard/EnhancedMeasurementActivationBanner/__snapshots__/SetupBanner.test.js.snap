// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupBanner should render correctly when the user does have the edit scope granted 1`] = `
<div>
  <section
    id="enhanced-measurement-notification"
  >
    <div
      class="googlesitekit-widget-context"
    >
      <div
        class="mdc-layout-grid"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
          >
            <div
              class="googlesitekit-banner googlesitekit-banner--setup-cta"
            >
              <div
                class="googlesitekit-banner__content"
              >
                <p
                  class="googlesitekit-typography googlesitekit-banner__title googlesitekit-typography--body googlesitekit-typography--medium"
                >
                  Understand how visitors interact with your content
                </p>
                <div
                  class="googlesitekit-banner__description"
                >
                  <span>
                    Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required.
                  </span>
                   
                  <a
                    aria-label="Learn more (opens in a new tab)"
                    class="googlesitekit-cta-link"
                    href="https://support.google.com/analytics/answer/9216061?hl=en-US"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    <span
                      class="googlesitekit-cta-link__contents"
                    >
                      Learn more
                    </span>
                    <span
                      class="googlesitekit-icon-wrapper"
                      style="margin-left: 5px;"
                    >
                      <svg />
                    </span>
                  </a>
                </div>
                <p
                  class="googlesitekit-typography googlesitekit-banner__help-text googlesitekit-typography--body googlesitekit-typography--medium"
                >
                  You can always add/edit this in the Site Kit Settings
                </p>
                <div
                  class="googlesitekit-notice__action"
                >
                  <button
                    class="mdc-button googlesitekit-banner__cta googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
                    target="_self"
                  >
                    <span
                      class="mdc-button__label"
                    >
                      Enable now
                    </span>
                  </button>
                  <button
                    class="mdc-button mdc-button--tertiary"
                    target="_self"
                  >
                    <span
                      class="mdc-button__label"
                    >
                      Maybe later
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="googlesitekit-banner__svg-wrapper googlesitekit-banner__svg-wrapper--center"
                style="background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PC9zdmc+);"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`SetupBanner should render correctly when the user does not have the edit scope granted 1`] = `
<div>
  <section
    id="enhanced-measurement-notification"
  >
    <div
      class="googlesitekit-widget-context"
    >
      <div
        class="mdc-layout-grid"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
          >
            <div
              class="googlesitekit-banner googlesitekit-banner--setup-cta"
            >
              <div
                class="googlesitekit-banner__content"
              >
                <p
                  class="googlesitekit-typography googlesitekit-banner__title googlesitekit-typography--body googlesitekit-typography--medium"
                >
                  Understand how visitors interact with your content
                </p>
                <div
                  class="googlesitekit-banner__description"
                >
                  <span>
                    Enable enhanced measurement in Analytics to automatically track metrics like file downloads, video plays, form interactions, etc. No extra code required — you’ll be redirected to give permission for Site Kit to enable it on your behalf.
                  </span>
                   
                  <a
                    aria-label="Learn more (opens in a new tab)"
                    class="googlesitekit-cta-link"
                    href="https://support.google.com/analytics/answer/9216061?hl=en-US"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    <span
                      class="googlesitekit-cta-link__contents"
                    >
                      Learn more
                    </span>
                    <span
                      class="googlesitekit-icon-wrapper"
                      style="margin-left: 5px;"
                    >
                      <svg />
                    </span>
                  </a>
                </div>
                <p
                  class="googlesitekit-typography googlesitekit-banner__help-text googlesitekit-typography--body googlesitekit-typography--medium"
                >
                  You can always add/edit this in the Site Kit Settings
                </p>
                <div
                  class="googlesitekit-notice__action"
                >
                  <button
                    class="mdc-button googlesitekit-banner__cta googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
                    target="_self"
                  >
                    <span
                      class="mdc-button__label"
                    >
                      Enable now
                    </span>
                  </button>
                  <button
                    class="mdc-button mdc-button--tertiary"
                    target="_self"
                  >
                    <span
                      class="mdc-button__label"
                    >
                      Maybe later
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="googlesitekit-banner__svg-wrapper googlesitekit-banner__svg-wrapper--center"
                style="background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PC9zdmc+);"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
